src/v/utils/retry_chain_node.cc

```diff
@@ -0,0 +1,298 @@
+    }
+    return nullptr;
+}
+
+const retry_chain_node* retry_chain_node::get_parent() const {
+    if (std::holds_alternative<retry_chain_node*>(_parent)) {
+        return std::get<retry_chain_node*>(_parent);
+    }
+    return nullptr;
+}
+
+ss::abort_source* retry_chain_node::get_abort_source() {
+    if (std::holds_alternative<ss::abort_source*>(_parent)) {
+        return std::get<ss::abort_source*>(_parent);
+    }
+    return nullptr;
+}
+
+const ss::abort_source* retry_chain_node::get_abort_source() const {
+    if (std::holds_alternative<ss::abort_source*>(_parent)) {
+        return std::get<ss::abort_source*>(_parent);
+    }
+    return nullptr;
+}
+
+retry_chain_node::~retry_chain_node() {
+    vassert(_num_children == 0, "Fiber stopped before its dependencies");
+    if (auto parent = get_parent(); parent != nullptr) {
+        parent->rem_child();
+    }
+}
+ss::sstring retry_chain_node::operator()() const { return format(); }
+
+retry_permit retry_chain_node::retry(retry_strategy st) {
+    auto as = find_abort_source();
+    if (as) {
+        as->check();
+    }
+    auto now = ss::lowres_clock::now();
+    if (
+      _deadline < now || _deadline == ss::lowres_clock::time_point::min()
+      || _retry == _max_retry_count) {
+        // deadline is not set or _retry counter is about to overflow (which
+        // will lead to 0ms backoff time) retries are not allowed
+        return {.is_allowed = false, .abort_source = as, .delay = 0ms};
+    }
+    auto required_delay = st == retry_strategy::backoff ? get_backoff()
+                                                        : get_poll_interval();
+    _retry++;
+    return {
+      .is_allowed = (now + required_delay) < _deadline,
+      .abort_source = as,
+      .delay = required_delay};
+}
+
+ss::lowres_clock::duration retry_chain_node::get_backoff() {
+    auto backoff = ss::lowres_clock::duration(_backoff * (1UL << _retry));
+    auto jitter = ss::lowres_clock::duration(
+      _retry_rand_jitter_source() % backoff.count());
+    return backoff + jitter;
+}
+
+ss::lowres_clock::duration retry_chain_node::get_poll_interval() {
+    auto jitter = _retry_rand_jitter_source() % _backoff;
+    return ss::lowres_clock::duration(
+      static_cast<ss::lowres_clock::duration::rep>(_backoff) + jitter);
+}
+
+ss::lowres_clock::duration retry_chain_node::get_timeout() {
+    auto now = ss::lowres_clock::now();
+    return now < _deadline ? _deadline - now : 0ms;
+}
+
+uint16_t retry_chain_node::get_len() const {
+    uint16_t len = 1;
+    auto next = get_parent();
+    while (next) {
+        len++;
+        next = next->get_parent();
+    }
+    return len;
+}
+
+ss::sstring retry_chain_node::format() const {
+    std::array<uint16_t, _max_retry_chain_depth> ids{_id};
+    int ids_len = 1;
+    auto next = get_parent();
+    while (next) {
+        ids.at(ids_len) = next->_id;
+        ids_len++;
+        next = next->get_parent();
+    }
+    std::stringstream str;
```

Comment:
fmt has a better string builder called `fmt::StringBuilder`  or smth like that. and you just `print_to` it like you did here.